Add static analysis for email-editor PHP package

lysyjan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes WOOPLUG-3234.

- This PR reintroduces the PHPStan static analysis command.
- The command should be executed as part of the CI job when PHP files in the email editor package are changed.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Checkout the branch
2. ~Install composer dependencies in the PHP email-editor package~
3. Execute ~`composer run phpstan` in `packages/php/email-editor` or~ `pnpm --filter='@woocommerce/email-editor-php-config' phpstan`
4. For testing on PHP 7.4 you can use `pnpm --filter='@woocommerce/email-editor-php-config' phpstan:php7` and for PHP 8.4 `pnpm --filter='@woocommerce/email-editor-php-config' phpstan` or `pnpm --filter='@woocommerce/email-editor-php-config' phpstan:php8`
5. The command `pnpm run phpstan` can be used directly in `packages/php/email-editor`
6. The command installs its own dependencies in `packages/js/email-editor/tasks/phpstan/vendor` with `stubs` packages. There is an optional parameter `--cleanup` removing those dependencies to avoid unwanted autocomplete in IDE.

<!-- End testing instructions -->

### Testing that has already taken place:

I tested the command locally with PHP 8.1.

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [x] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

lysyjan committed:
Add phpstan dependecies and create config

lysyjan committed:
Update phpstan configuration and add ignored errros

There is a huge amount of error that can be difficult to fix. Some of them can be addressed later in this PR.

lysyjan committed:
Fix PHPStan error

Fixing those errors should make the code safer.

lysyjan committed:
Prepare PHPStan for compatibility with tests directory

lysyjan committed:
Fix PHPStan errors in tests directory

lysyjan committed:
Update PHPStan commands and add paths to config

lysyjan committed:
Add changelog file

costasovo commented on packages/php/email-editor/composer.json:
```diff
@@ -47,7 +47,7 @@
 		"phpcs": "phpcs -s -p",
 		"phpcbf": "phpcbf -p",
-		"phpstan": "php ./tasks/run-phpstan.php"
+		"phpstan": "vendor/bin/phpstan analyse -c phpstan.neon"
```
I would suggest we add a parameter for memory to the command

```
"phpstan": "vendor/bin/phpstan analyse -c phpstan.neon --memory-limit=2G"
```

This will avoid memory limit errors when someone has PHP configured with a strict memory limit locally.

costasovo reviewed (commented):
@lysyjan Thanks, you for adding the analysis and for fixing the issues. Great job!

There is an issue with the core and woo stubs I think we need to address. Please see my other comment.

I was clicking through the CI jobs below this PR, but was not able to find where the PHPStan ran. Could you please point me to the job? I assumed it there should be a run because PHP files changed.

What PHP version will be used in the CI job?

I tested it locally with phpVersions from 7.4 to 8.4 set in the phpstan.neon. All were green. Only the 8.4 reported two errors, but we don't need to address those now:
```
21     Deprecated in PHP 8.4: Parameter #3 $previous (Throwable) is implicitly nullable via default value null.  
         🪪 parameter.implicitlyNullable                                                                           
  26     Deprecated in PHP 8.4: Parameter #1 $previous (Throwable) is implicitly nullable via default value null.  
         🪪 parameter.implicitlyNullable  
```

lysyjan committed:
Add memory parameter for PHPStan

lysyjan committed:
Update CI config in package.json for email-editor

lysyjan committed:
Add support for PHPStan for PHP 7 and 8

lysyjan committed:
Delete outdated file for running PHPStan

lysyjan committed:
Fix errors related for PHPStan with PHP 8.4

lysyjan commented:
@costasovo I fixed the mentioned issues.

I updated the command names and tried to use the most straightforward approach for running PHPStan with different PHP versions. Let me know if you prefer a different solution.

I also needed to make minor adjustments in the README file because the markdown validation job failed.

costasovo commented:
@lysyjan Thanks for the update! I can now see the jobs below the PR!

There is still the issue with stubs. I can't see my comment I was writing yesterday so it looks like I forgot to submit that particular one 🤦 Sorry, my bad!

The problem is that the Woo and WordPress stubs we install for PHPStan get picked by IDEs and it causes for example, that they are offered by the IDE to developers when they go to the function definition. See screenshot where I try to click get to  `getSubject` source. Instead of going directly to the source, I get these three options to choose from.

<img width="1237" alt="Screenshot 2025-05-22 at 7 37 02" src="https://github.com/user-attachments/assets/10874f5b-301b-433f-90fa-90347e27041e" />

I'm afraid we can't merge it like this because this will affect the dev experience for all users of the monorepo. I'm not sure what is the best solution. I think we could install/unzip? these stubs only for the job and then remove them. Or perhaps there might be another way to prevent IDEs from picking them up.

lysyjan committed:
Create a separate directory for PHPStan command

- Because we want to avoid including stubs packages in IDE's autocomplete, we decided to create a separated directory with required dependencies.
- There is a bash script for running PHPStan for better readability and easier logic management.

lysyjan committed:
Exclude PHPStan temp directory from CodeSniffer config

lysyjan commented:
@costasovo Thank you. I added two new commits and rebased the branch.
Please test if the commands work as expected.

I decided to create a simple bash script for better readability.

costasovo reviewed (commented):
@lysyjan Thanks for the updates! It works nicely!

I see I need to use `pnpm phpstan:php7 --cleanup` to run it. So you may need to update README.md which says composer phpstan...

I suggest making the `--cleanup` argument default behavior and perhaps changing the logic to `--skip-cleanup`. 
The files are locally installed from composer cache, so I don't observe any difference in the speed, and I think it might be better for someone who would work with the package only occasionally and may forget or just don't know that they need to clean up.
But it is only a suggestion. I'm also fine with keeping it as is as I can simply use it with the argument.

lysyjan committed:
Change the optional argument from cleanup to skip-cleanup

- The default PHPStan directory cleanup should offer better developer experience.
- The commit contains also updating the readme file.

lysyjan commented:
@Konamiman Thank you for testing. I need to apologize because I forgot to update the testing instructions in the description. After some @costasovo suggestions, we decided to create a separate `composer.json`. The main reason was loading the stubs packages by IDEs. More details are [here](https://github.com/woocommerce/woocommerce/pull/58135#issuecomment-2899991597).

> This seems to work (progress bar reaches 100%) but at the end it fails with "Internal error: This enum case does not have a value while analysing file"

If you run PHPStan on PHP 7.4, you should use the command `phpstan:php7`. I updated the testing instructions and the package readme file.

@costasovo I updated the readme file and changed the option `cleanup` to `skip-cleanup`
